extensions-contrib/materialized-view-selection/src/main/java/io/druid/query/materializedview/DatasourceOptimizer.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.materializedview;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSortedSet;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.inject.Inject;
+import io.druid.client.TimelineServerView;
+import io.druid.client.selector.ServerSelector;
+import io.druid.query.Query;
+import io.druid.query.TableDataSource;
+import io.druid.query.groupby.GroupByQuery;
+import io.druid.query.spec.MultipleIntervalSegmentSpec;
+import io.druid.query.timeseries.TimeseriesQuery;
+import io.druid.query.topn.TopNQuery;
+import io.druid.timeline.TimelineObjectHolder;
+import org.joda.time.Interval;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Consumer;
+
+public class DatasourceOptimizer 
+{
+  private static final ReadWriteLock lock = new ReentrantReadWriteLock();
+  private static ConcurrentHashMap<Derivative, AtomicLong> derivativesHitCount = new ConcurrentHashMap<>();
+  private static ConcurrentHashMap<String, AtomicLong> totalCount = new ConcurrentHashMap<>();
+  private static ConcurrentHashMap<String, AtomicLong> hitCount = new ConcurrentHashMap<>();
+  private static ConcurrentHashMap<String, AtomicLong> costTime = new ConcurrentHashMap<>();
+  private static ConcurrentHashMap<String, ConcurrentHashMap<Set<String>, AtomicLong>> missFields = new ConcurrentHashMap<>();
+  private static TimelineServerView serverView = null;
```

Comment:
`serverView` is used in `optimize` method, and this method is static.

Reply:
I mean, this is should be a final non-static variable because it's quite dangerous. As you said, `serverView` is used in a static method (`optimize()`), but is initialized in the constructor. As you know, static methods can be used without creating an instance which means `serverView` might not be initialized when `optimize()` is called. This currently works because Guice initializes `DataSourceOptimizer` when `DataSourceOptimizerMonitor` is initialized and this happens to be before `optimize()` is called. However, it might be broken in the future if somethings change like someone decides to make `DataSourceOptimizerMonitor` configurable and disables it.

Reply:
Thanks, you'r right. I'll modify it